derp/derp_server.go

```diff
@@ -424,51 +423,24 @@ type sclientInfo struct {
 type serverInfo struct {
 }
 
-// Stats returns stats about the server.
-func (s *Server) Stats() *ServerStats {
-	s.mu.Lock()
-	defer s.mu.Unlock()
-	return &ServerStats{
-		BytesPerSecondLimit: s.BytesPerSecond,
-		CurrentConnections:  len(s.netConns),
-		UniqueClientsEver:   len(s.clientsEver),
-		TotalAccepts:        s.accepts,
-		BytesReceived:       atomic.LoadInt64(&s.bytesRecv),
-		BytesSent:           atomic.LoadInt64(&s.bytesSent),
-		PacketsDropped:      atomic.LoadInt64(&s.packetsDropped),
-		PacketsReceived:     atomic.LoadInt64(&s.packetsRecv),
-		PacketsSent:         atomic.LoadInt64(&s.packetsSent),
-	}
+func (s *Server) expVarFunc(f func() interface{}) expvar.Func {
+	return expvar.Func(func() interface{} {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		return f()
+	})
 }
 
 // ExpVar returns an expvar variable suitable for registering with expvar.Publish.
 func (s *Server) ExpVar() expvar.Var {
-	return expVar{s}
-}
-
-type expVar struct{ *Server }
-
-// String implements the expvar.Var interface, returning the current server stats as JSON.
-func (v expVar) String() string {
-	ss := v.Server.Stats()
-	j, err := json.MarshalIndent(ss, "", "\t")
-	if err != nil {
-		return "{}"
-	}
-	return string(j)
-}
-
-// ServerStats are returned by Server.Stats.
-//
-// It is JSON-ified by expVar for the expvar package.
-type ServerStats struct {
-	BytesPerSecondLimit int   `json:"bytesPerSecondLimit"`
-	CurrentConnections  int   `json:"currentClients"`
-	UniqueClientsEver   int   `json:"uniqueClientsEver"`
-	TotalAccepts        int64 `json:"totalAccepts"`
-	BytesReceived       int64 `json:"bytesReceived"`
-	BytesSent           int64 `json:"bytesSent"`
-	PacketsDropped      int64 `json:"packetsDropped"`
-	PacketsReceived     int64 `json:"packetsReceived"`
-	PacketsSent         int64 `json:"packetsSent"`
+	m := new(expvar.Map)
+	m.Set("gauge_current_connnections", s.expVarFunc(func() interface{} { return len(s.netConns) }))
+	m.Set("counter_unique_clients_ever", s.expVarFunc(func() interface{} { return len(s.clientsEver) }))
+	m.Set("accepts", &s.accepts)
```

Comment:
Missing type hints on all the following (our exporter should fail/complain/not export variables that don't specify a type)